aiohttp/web_scheduler.py

```diff
@@ -0,0 +1,226 @@
+                context['source_traceback'] = self._source_traceback
+            scheduler.call_exception_handler(context)
+
+    def _done_callback(self, task):
+        scheduler = self._scheduler
+        scheduler._done(self, False)
+        try:
+            exc = task.exception()
+        except asyncio.CancelledError:
+            pass
+        else:
+            if exc is not None and not self._explicit_wait:
+                context = {'message': "Job processing failed",
+                           'job': self,
+                           'exception': exc}
+                if self._source_traceback is not None:
+                    context['source_traceback'] = self._source_traceback
+                scheduler.call_exception_handler(context)
+                scheduler._failed_tasks.put_nowait(task)
+        self._runner = None  # drop backref
+        self._closed = True
+
+
+class Scheduler(Container):
+    def __init__(self, *, loop):
+        self._loop = loop
+        self._jobs = set()
+        self._close_timeout = 0.1
+        self._concurrency = 100
+        self._exception_handler = None
+        self._failed_tasks = asyncio.Queue(loop=loop)
+        self._failed_waiter = ensure_future(self._wait_failed(), loop=loop)
+        self._pending = deque()
+        self._closed = False
+
+    @asyncio.coroutine
+    def run(self, coro):
+        job = Job(coro, self, self._loop)
+        if self._concurrency is None or self.active_count < self._concurrency:
+            job._start()
+        else:
+            self._pending.append(job)
+        self._jobs.add(job)
+        return job
+
+    def __iter__(self):
+        return iter(list(self._jobs))
+
+    def __len__(self):
+        return len(self._jobs)
+
+    def __contains__(self, job):
+        return job in self._jobs
+
+    def __repr__(self):
+        info = []
+        if self._closed:
+            info.append('closed')
+        info = ' '.join(info)
+        if info:
+            info += ' '
+        return '<Scheduler {}jobs={}>'.format(info, len(self))
+
+    @property
+    def closed(self):
+        return self._closed
+
+    @asyncio.coroutine
+    def close(self):
+        if self._closed:
+            return
+        jobs = self._jobs
+        if jobs:
+            yield from asyncio.gather(*[job.close() for job in jobs],
+                                      loop=self._loop, return_exceptions=True)
+        self._pending.clear()
+        self._jobs.clear()
+        self._failed_tasks.put_nowait(None)
+        yield from self._failed_waiter
+        self._closed = True
+
+    @property
+    def concurrency(self):
+        return self._concurrency
+
+    @concurrency.setter
+    def concurrency(self, concurrency):
+        self._concurrency = concurrency
+
+    @property
+    def active_count(self):
+        return len(self._jobs) - len(self._pending)
+
+    @property
+    def pending_count(self):
+        return len(self._pending)
```

Comment:
Just a question: how about to return active/pending list of tasks instead of their count? Then user may decide what to do with them: count them all, close them all, filter out to find out something.